### ISSUE TITLE
IOS Memory Leak

### ISSUE THREAD

**Issue by kristiantoska** [TheWidlarzGroup/react-native-video#3399]
# Bug

## Platform
Which player are you experiencing the problem on:
* iOS

## Environment info

Library version: 6.0.0-beta.0
Device: iPhone 14 Pro Max (Emulator)

## Steps To Reproduce

1. Show video component with an online video (in the example i am using the following 30mb sample video: https://sample-videos.com/video123/mp4/480/big_buck_bunny_480p_30mb.mp4)
2. Unmount the video component
3. Re-Mount the component
4. The Memory usage of the app grows by about the size of the video in my case, and keeps growing with each repeat
5. Profiling also seems to report several memory leaks, with most of the extra memory allocation going into `VM: CoreMedia HTTP cache`
<img width="540" alt="Screenshot 2023-12-04 at 19 25 17" src="https://github.com/react-native-video/react-native-video/assets/20017227/ca1a31e4-6a99-4da8-8acf-c916ddd8d54b">

## Expected behaviour

1. Video data is cleared from memory whenever the `Video` component is unmounted

## Reproducible sample code

Full repo [here](https://github.com/kristiantoska/RNVideo-Memory-Leak-Test): Clean RN init + RNVideo@6.0.0-beta.0 install only

```
import React, {useState} from 'react';
import {SafeAreaView, TouchableOpacity, Text, View} from 'react-native';
import Video from 'react-native-video';

function App() {
  const [hidden, setHidden] = useState(false);

  return (
    <SafeAreaView style={{flex: 1}}>
      <View style={{height: 300, width: '100%'}}>
        {!hidden && (
          <Video
            style={{height: 300, width: '100%'}}
            source={{
              uri: 'https://sample-videos.com/video123/mp4/480/big_buck_bunny_480p_30mb.mp4',
            }}
            controls
            paused={false}
          />
        )}
      </View>

      <TouchableOpacity
        style={{padding: 20, backgroundColor: hidden ? 'red' : 'green'}}
        onPress={() => setHidden(!hidden)}>
        <Text>Video: {hidden ? 'Unmounted' : 'Mounted'}</Text>
      </TouchableOpacity>
    </SafeAreaView>
  );
}

export default App;

```

## Video sample

https://github.com/react-native-video/react-native-video/assets/20017227/beecff73-7bdf-4acf-a70e-c3d6ea49286a

**Comment by unubold1** [comment]
yes i worry about that too. How to fix this memory leak.

**Comment by Nikualla1** [comment]
Happening to me too. :/ +1

**Comment by andybui2025** [comment]
The same with FeedScreen using FlatList, try alpha 7 version work perfect, waiting for new update

**Comment by DongGukMon** [comment]
I have the same problem. I even checked that memory is released in version 5.2.1. I was wondering if deinit is not being called, so I tried calling deinit method directly from react native with nativeModule, but memory leak was still occurring. I'm going to downgrade to 5.2.1 and use it until the problem is solved.🫠

**Comment by amserhii** [comment]
Same issue. 5.2.1 and alpha 7 does not work. using 6.0.0-beta 3.

**Comment by WriteCodeEveryday** [comment]
Having issues with this as well on 5.2.1 and 6.0.0-beta.3

**Comment by james-cohen** [comment]
Have had the same issue across multiple versions including 6.0.0-beta.3. Please can this be prioritised for fixing ASAP, we're having to do loads of weird hacks to disguise the video player crashing after a few videos

**Comment by amserhii** [comment]
Ok, I have fixed a lib a little, and now deinit method is called in RCTVideo, btw, it does not help and memory still increasing a lot.

`deinit {
        NSLog("DEINITED")
        NotificationCenter.default.removeObserver(self)
        self.removePlayerLayer()
        self._player?.replaceCurrentItem(with: nil)
        _playerObserver.clearPlayer()
         _player = nil
        _playerItem = nil
        _videoURL = nil
        URLCache.shared.removeAllCachedResponses()
    }`

**Comment by kristiantoska** [comment]
After looking into this for a while i noticed that the issue is caused by the PIP IOS support. I managed to fix the issue on the latest releases by commenting this line on native as I was not using PIP: 
https://github.com/react-native-video/react-native-video/blob/01d7bedb4160a2be55708d45721eb7b12407519d/ios/Video/RCTVideo.swift#L143

**Comment by amserhii** [comment]
@kristiantoska  could you please provide how do you render the list of items? Because I just tried your changes and it does not help. May be you can provide us with part of your RCTVideo.swift and your changes. 
I am using FlatList to render items and un-mount it on condition inside list. Looks like memory released but not fully. If I am scrolling long it still fill our RAM

**Comment by alejandrorbradford** [comment]
Same issue here. Looks like the component isn't getting released. Any ideas? When investigating, I noticed that on the delegate methods when app enters foreground/background, the observers are also not getting removed on `RCTVideo`

**Comment by amserhii** [comment]
@alejandrorbradford  Which one of methods not getting removed?

**Comment by amserhiiio** [comment]
Any updates, guys? Critical issue and still no reply

**Comment by amserhii** [comment]
Any updates guys? So critical issue and no replies

**Comment by james-cohen** [comment]
I'm assuming this is also related to [this issue](https://github.com/react-native-video/react-native-video/issues/1938) that was seemingly never resolved?

**Comment by kristiantoska** [comment]
@amserhii In my case i was not rendering a list of videos, just playing one video on screen whenever the user opened an article, and the `pip` hack i mentioned above improved the memory leak by a lot for me, but it is still very slightly present

**Comment by james-cohen** [comment]
> After looking into this for a while i noticed that the issue is caused by the PIP IOS support. I managed to fix the issue on the latest releases by commenting this line on native as I was not using PIP:
> 
> https://github.com/react-native-video/react-native-video/blob/01d7bedb4160a2be55708d45721eb7b12407519d/ios/Video/RCTVideo.swift#L143

I can confirm that commenting this out fixed the issue for me, thank you @kristiantoska

**Comment by KrzysztofMoch** [comment]
Should be fixed with `6.0.0-beta.5`, feel free to repo if it doesn't

**Comment by murilogok** [comment]
> Should be fixed with `6.0.0-beta.5`, feel free to repo if it doesn't

It worked, tha
![Screenshot 2024-02-28 at 03 44 44](https://github.com/react-native-video/react-native-video/assets/83665217/e95d45f0-7771-496f-a1f0-1e8388eb9e1d)
nks a lot

**Comment by murilogok** [comment]
> Should be fixed with `6.0.0-beta.5`, feel free to repo if it doesn't

It works! Thanks a lot

![Screenshot 2024-02-28 at 03 44 44](https://github.com/react-native-video/react-native-video/assets/83665217/dc070481-eb83-4d95-9927-47ec63bd4740)
